src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnectionImpl.java

```diff
@@ -1292,10 +1298,17 @@ public NotificationResult run() {
                         return getServerNotifFwd().fetchNotifs(csn, t, mn);
                     }
             };
-            if (acc == null)
-                return action.run();
-            else
+            if (acc == null) {
+                // No ACC, therefore no SM. May have a Subject:
+                if (subject != null) {
+                    return Subject.doAs(subject, action);
```

Comment:
Is it ever possible for subject to be `null` (passed in as `null` to the constructor) and an SM to be enabled? If so, then the call above to `Subject.doAs` would trigger a permission for an `AuthPermission("doAs")` permission. (This also may have been the case you were seeing in tests).

I think following Weijun's advice above is cleaner and safer, so you do one or the other depending on the allowSM setting, and not whether certain variables are null or not.

Reply:
Right, the only possible assignment to acc in this file is if we were given a Subject, and SM is permitted.

In future there will be a Subject, which can be null.  While we handle SM, we still use the ACC if RMIConnectionImpl was created with a Subject.